### PR TITLE
[VectorDistribution] Remove iteration on non-distributed levels

### PR DESCRIPTION
**Pull request by Groverkss**
This patch removes iteration order in the layout for non-distributed levels. The layout now only represents how the register is tiled, and does not define an iteration order over it. The patterns themselves define an iteration order over the tiled layout.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]:
```suggestion
  ArrayRef<int64_t> batchOffsets(offsets.begin(), rank);
```

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]:
```suggestion
  ArrayRef<int64_t> outerVectorOffsets(offsets.begin() + rank, rank);
```

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp:595]: To be crystal clear, you can mention that each thread **(0-3)** carries the same data.
(IE clarify that the indices represent thread id instead of the data)

**Review by kuhar — APPROVED** [review]
Nice cleanup! Please wait for a review from one more person before landing (e.g., Quinn)

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: TIL

**Comment by Groverkss** [comment]
Matmul test suite passes on MI250 with this patch

File: compiler/src/iree/compiler/Codegen/Common/GPU/AMDGPUDistributeContract.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/AMDGPUDistributeContract.cpp:217]: Just noting a bug here: I think this is wrong if the order of M/N dims in the inputs does not match the order of those dims in the C matrix.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]:
```suggestion
    // The distributed reduction mask is simply the same mask appended
```

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp:634]: dropped the wrong errs print here (subgroup basis instead of element orders)

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp:683]: same here

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp:675]: same here

**Review by qedawkins — APPROVED** [review]
Approving because I think this is probably worth doing, but it will be a lot harder to get back in the future if needed, so I would feel much better with some confirmation of performance of SDXL with it, specifically because it changes the unroll order of vector transfers. I'd hope it doesn't really matter though.

File: compiler/src/iree/compiler/Codegen/Common/GPU/AMDGPUDistributeContract.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/GPU/AMDGPUDistributeContract.cpp:217]: (bug was already there before, don't need to fix anything)

**Pull request merged**
